### PR TITLE
Format StmtFor

### PR DESCRIPTION
**Pull request by davidszotten**
## Summary

format StmtFor

still trying to learn how to help out with the formatter. trying something slightly more advanced than [break](#5158)

mostly copied form StmtWhile

## Test Plan

snapshots

### CODE REVIEW

File: crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__ruff_test__statement__for_py.snap
Comment (konstin) [inline crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__ruff_test__statement__for_py.snap:60]: black doesn't break this for statement, but i think this actually makes more sense

File: crates/ruff_python_formatter/src/statement/stmt_for.rs
Comment (konstin) [inline crates/ruff_python_formatter/src/statement/stmt_for.rs:38]: We might need a new setting that actually removes expression parentheses if not used, i'll experiment with this

**Review by konstin — COMMENTED** [review]
You picked a tough node but this looks good barring ExprTuple always keeping its parentheses (because i missed that case in ExprTuple)

File: crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__ruff_test__statement__for_py.snap
Comment (davidszotten) [reply crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__ruff_test__statement__for_py.snap:60]: hm. interesting (how did you notice the difference)

at this point, do we have a policy/guidelines for matching black (to help people migrate) vs improving formatting? maybe we raise an issue or otherwise to see if the black maintainers also prefer this formatting?

**Comment by davidszotten** [comment]
> You picked a though node

Impl is mostly copied from `StmtWhile` which made it easier. Still trying to learn my way around

File: crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__ruff_test__statement__for_py.snap
Comment (konstin) [reply crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__ruff_test__statement__for_py.snap:60]: Our goal is to match black except in edge cases, but we haven't formalized what exactly this means. I'll discuss with @MichaReiser tomorrow what we want do about this

> how did you notice the difference?

copied the code and ran black over it ;)

Comment (konstin) [reply crates/ruff_python_formatter/src/snapshots/ruff_python_formatter__tests__ruff_test__statement__for_py.snap:60]: at the moment, yes

File: crates/ruff_python_formatter/src/statement/stmt_for.rs
Comment (davidszotten) [reply crates/ruff_python_formatter/src/statement/stmt_for.rs:38]: do we want to keep this open pending that or merge as is for now?

**Comment by konstin** [comment]
Finally got around to it: https://github.com/astral-sh/ruff/pull/5175 

This should hopefully unblock you, using
```rust
#[derive(Debug)]
struct ExprTupleWithoutParentheses<'a>(&'a Expr);

impl Format<PyFormatContext<'_>> for ExprTupleWithoutParentheses<'_> {
    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
        match self.0 {
            Expr::Tuple(expr_tuple) => expr_tuple
                .format()
                .with_options(TupleParentheses::StripInsideForLoop)
                .fmt(f),
            other => other.format().with_options(Parenthesize::IfBreaks).fmt(f),
        }
    }
}
```
and `ExprTupleWithoutParentheses(target.as_ref()),`

File: crates/ruff_python_formatter/src/statement/stmt_for.rs
Comment (konstin) [inline crates/ruff_python_formatter/src/statement/stmt_for.rs:66]: Could you add a `debug_assert!` here that in the `orelse.is_empty()` case `or_else_comments` is also empty?

**Review by konstin — APPROVED** [review]

**Review by konstin — APPROVED** [review]
thanks

**Pull request merged**
